Analyzer driver performance improvements

mavasani opened the pull request:
This change improves the analyzer execution performance by plugging in an optional semantic model provider into the compilation. This allows analyzer callbacks to fully share semantic model and bound nodes with the compiler binding phase. Additionally, analyzers invoking `compilation.GetSemanticModel` also get the shared semantic model, further improving performance. Analyzer driver ensures that we do not hold onto the cached semantic model/bound nodes for the entire lifetime of the compilation by dropping the cached semantic model after processing the `CompilationUnitCompletedEvent` for each syntax tree, which indicates the entire syntax tree's binding and analysis phase has completed.

Performance measurements on the compiler benchmark suite for analyzer execution and real world repo shows ~10-15% improvement in analyzer execution time. There is also a large allocation improvement (~30%) in analyzer execution, though majority of this chunk comes from caching some delegates that were being created on hot paths, rather then from sharing semantic model through this provider.

**Strongly recommended to review commit-by-commit**
1. Commit 1 (https://github.com/dotnet/roslyn/commit/f06d87c7c62a8785af74497c119e55fd0b64ef14): Nullable enable analyzer driver. I also have a separate PR out for this commit, if that is easier to review: https://github.com/dotnet/roslyn/pull/46409

2. Commit 2 (https://github.com/dotnet/roslyn/commit/039c424a0a582b4730729f40d5824e7220967087): Semantic model provider implementation for the performance improvements

## Performance numbers

### 1. Compiler benchmark

Performance suite: https://github.com/dotnet/roslyn/blob/master/docs/wiki/Measuring-Compiler-Performance.md

#### Master branch

|                      Method |    Mean |   Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |--------:|--------:|--------:|--------:|--------:|--------:|------------:|-----------:|----------:|----------:|
| GetDiagnosticsWithAnalyzers | **13.90 s** | 0.139 s | 0.190 s | 13.87 s | 13.45 s | 14.37 s | 318000.0000 | 85000.0000 | 2000.0000 |   **1.85 GB** |

#### This PR branch

|                      Method |    Mean |   Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |--------:|--------:|--------:|--------:|--------:|--------:|------------:|-----------:|----------:|----------:|
| GetDiagnosticsWithAnalyzers | **12.51 s** | 0.124 s | 0.122 s | 12.53 s | 12.33 s | 12.73 s | 216000.0000 | 63000.0000 | 2000.0000 |   **1.27 GB** |

### 2. Building real world repo with analyzers

Scenario: Rebuild of RoslynAnalyzers.sln https://github.com/dotnet/roslyn-analyzers repo

1. Master branch: 72 seconds
2. This PR branch: 61 seconds

mavasani commented:
FYI @sharwell @CyrusNajmabadi 

@AlekseyTs - this is the performance optimization in analyzer driver that we talked about a while back. We now hook up a semantic model provider to the compilation when executing analyzers, such that the analyzer driver controls the lifetime and caching of the shared semantic model with bound nodes.

mavasani commented:
Also tagging @gafter - this change drops the weak reference to semantic model with cached bound nodes in `SymbolDeclaredCompilationEvent` in favor of a semantic model provider that has strong reference to semantic model with cached bound nodes, such that analyzer driver controls its lifetime.

CyrusNajmabadi commented:
i don't know this subsystem well enough to review.  but i strongly approve of us doing something here.  teh design seems sane, and i think it's good that it becomes an explicit 'owned' things by a provider, rather than an impl detail that depends on weak-refs.

AlekseyTs commented:
>         public new SemanticModel GetSemanticModel(SyntaxTree syntaxTree, bool ignoreAccessibility)

It looks like we are missing an overload for this API. I assume for VB as well.

---
Refers to: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2105 in f087408. [](commit_id = f087408eb126b198e6295b04ed0b902aec05ce1a, deletion_comment = False)

mavasani commented:
> It looks like we are missing an overload for this API. I assume for VB as well.

Thanks @AlekseyTs. I have addressed the feedback and also added unit tests for the same.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -501,6 +506,11 @@ public bool ContainsSyntaxTree(SyntaxTree syntaxTree)
+        /// <summary>
+        /// Optional semantic model provider for this compilation.
+        /// </summary>
+        internal readonly SemanticModelProvider? SemanticModelProvider;
```
>internal [](start = 8, length = 8)

We don't want to expose it to the public?

AlekseyTs replied:
I see, it is itself internal

---
In reply to: [465317125](https://github.com/dotnet/roslyn/pull/46508#discussion_r465317125) [](ancestors = 465317125)

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/SemanticModelProvider.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract class SemanticModelProvider
```
>SemanticModelProvider [](start = 28, length = 21)

Consider adding doc comment. Is there a common usage pattern or main consumer?

mavasani replied:
Will add doc comment, the primary consumer right now is analyzer driver or any place that was caching semantic models for tree for known performance reasons (for example SuppressMessageAttributeState.IsDiagnosticSuppressed)

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/SemanticModelProvider.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    internal abstract class SemanticModelProvider
+    {
+        public abstract SemanticModel GetSemanticModel(SyntaxTree tree, Compilation compilation);
```
>GetSemanticModel [](start = 38, length = 16)

Doc comment?

AlekseyTs commented on src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs:
```diff
@@ -343,11 +343,15 @@ internal static Diagnostic Create(DiagnosticInfo info)
 
+            if (compilation.SemanticModelProvider == null)
+            {
+                compilation = compilation.WithSemanticModelProvider(new CachingSemanticModelProvider());
```
>new CachingSemanticModelProvider() [](start = 68, length = 34)

What are we trying to achieve by this and why? Consider explaining this in a comment.

mavasani replied:
I will revert this change - I originally thought this was a hot path, but this is actually a separate public API that is not used during analyzer execution.

AlekseyTs commented on src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs:
```diff
@@ -343,11 +343,15 @@ internal static Diagnostic Create(DiagnosticInfo info)
 
+            if (compilation.SemanticModelProvider == null)
+            {
+                compilation = compilation.WithSemanticModelProvider(new CachingSemanticModelProvider());
```
>compilation [](start = 16, length = 11)

It doesn't feel right to use a different compilation in SuppressMessageAttributeState

AlekseyTs commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs:
```diff
@@ -0,0 +1,40 @@
+
+        public CachingSemanticModelProvider()
+        {
+            _semanticModelsMap = new ConcurrentDictionary<SyntaxTree, SemanticModel>();
```
>ConcurrentDictionary [](start = 37, length = 20)

What is the lifetime of this dictionary? It looks like it is going to keep added trees alive for as long as it is alive.

mavasani replied:
The dictionary will stay alive for the lifetime of the provider, but the entries will be removed from the map by the analyzer driver whenever analysis completes for each tree.

mavasani replied:
I have added a doc comment to the type for clarity.

AlekseyTs commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs:
```diff
@@ -0,0 +1,40 @@
+        public override SemanticModel GetSemanticModel(SyntaxTree tree, Compilation compilation)
+        {
+            if (_semanticModelsMap.TryGetValue(tree, out var model) &&
+                model.Compilation == compilation)
```
>model.Compilation == compilation [](start = 16, length = 32)

Should we simply use compilation as part of the key?

mavasani replied:
That seems reasonable, let me make that change.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -633,7 +641,27 @@ public CSharpCompilation WithScriptCompilationInfo(CSharpScriptCompilationInfo?
+        /// </summary>
+        internal override Compilation WithSemanticModelProvider(SemanticModelProvider? semanticModelProvider)
+        {
+            return new CSharpCompilation(
```
>return new CSharpCompilation( [](start = 12, length = 29)

Missing a check to see if we already using this provider?

mavasani replied:
Yes, sure.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2085,7 +2114,28 @@ public new SemanticModel GetSemanticModel(SyntaxTree syntaxTree, bool ignoreAcce
                 throw new ArgumentException(CSharpResources.SyntaxTreeNotFound, nameof(syntaxTree));
             }
```
Are these checks redundant now?

mavasani replied:
Yes, thanks. Will remove.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -683,7 +691,30 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+        ''' Returns a new compilation with the given semantic model provider.
+        ''' </summary>
+        Friend Overrides Function WithSemanticModelProvider(semanticModelProvider As SemanticModelProvider) As Compilation
+            Return New VisualBasicCompilation(
```
>Return New VisualBasicCompilation( [](start = 12, length = 34)

Same question as for C#

mavasani replied:
Yes, thanks.

mavasani commented:
@AlekseyTs any further feedback?

@dotnet/roslyn-compiler for additional review. Thanks.

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs:
```diff
@@ -0,0 +1,46 @@
+            }
+
+            // Avoid infinite recursion by passing 'useSemanticModelProviderIfNonNull: false'
+            model = compilation.GetSemanticModelCore(tree, ignoreAccessibility: false, useSemanticModelProviderIfNonNull: false);
```
Is there a reason we don't make the `ignoreAccessibility` configurable by the caller?

mavasani replied:
We can make this configurable, it just wasn't required for any internal implementations of semantic model provider.

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs:
```diff
@@ -0,0 +1,46 @@
+            return _semanticModelsMap.GetOrAdd((tree, compilation), model);
+        }
+
+        internal void RemoveCachedSemanticModel(SyntaxTree tree, Compilation compilation)
```
The name doesn't quite match up with the behavior here. Consider: `EnsureCachedSemanticModelRemoved` or change this method to return `bool` and return the result.

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs:
```diff
@@ -0,0 +1,46 @@
+
+        public override SemanticModel GetSemanticModel(SyntaxTree tree, Compilation compilation)
+        {
+            if (_semanticModelsMap.TryGetValue((tree, compilation), out var model))
```
This method is going to essentially do the double lookup pattern: try then get. Is there a reason you didn't use the `GetOrAdd` overload that takes a `Func<SemanticModel>` to avoid the double lookup? Believe this can be done to avoid extra allocations.

mavasani replied:
Originally I did not have compilation as part of the key, so I had used this pattern, but now that compilation is also part of the key, we can switch to single lookup approach.

jaredpar commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -501,6 +506,11 @@ public bool ContainsSyntaxTree(SyntaxTree syntaxTree)
+        /// <summary>
+        /// Optional semantic model provider for this compilation.
+        /// </summary>
+        internal readonly SemanticModelProvider? SemanticModelProvider;
```
Why expose this as a field instead of a property?

mavasani replied:
Because I blatantly copy pasted the previous member: https://github.com/dotnet/roslyn/blob/789a25bbb89a48e964d3b0fa951d02f72d36ee43/src/Compilers/Core/Portable/Compilation/Compilation.cs#L504-L507

I will switch both of these to be auto-props.

jaredpar commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -203,11 +210,9 @@ public Compilation Clone()
+            => GetSemanticModelCore(syntaxTree, ignoreAccessibility, useSemanticModelProviderIfNonNull: true);
 
-        protected abstract SemanticModel CommonGetSemanticModel(SyntaxTree syntaxTree, bool ignoreAccessibility);
+        internal abstract SemanticModel GetSemanticModelCore(SyntaxTree syntaxTree, bool ignoreAccessibility, bool useSemanticModelProviderIfNonNull);
```
Think it would be clearer to have two methods here:

1. `GetSemanticModelCore`: this gets a `SemanticModel` which is potentially a cached instance. It's at the discretion of the `Compilation` to make the decision on how it is created and returned
1. `CreateSemanticModelCore`: this always creates a new `SemanticModel`. I find that is much clearer than calling `Get` and using a flag `useSemanticModelProviderIfNonNull: false`.

mavasani replied:
Ah, that is much better, will split.

jaredpar commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2085,7 +2122,13 @@ public new SemanticModel GetSemanticModel(SyntaxTree syntaxTree, bool ignoreAcce
-            return new SyntaxTreeSemanticModel(this, (SyntaxTree)syntaxTree, ignoreAccessibility);
+            if (SemanticModelProvider != null && useSemanticModelProviderIfNonNull)
+            {
+                Debug.Assert(!ignoreAccessibility);
```
This is another reason to prefer two methods: `GetSemanticModelCore` and `CreateSemanticModel`. The input to the single method is only conditionally correct. Removes an entire class of errors if we have two separate methods here.

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:
```diff
@@ -1281,30 +1283,23 @@ private static IEnumerable<Diagnostic> GetEffectiveDiagnosticsImpl(ImmutableArra
                 yield break;
             }
 
+            if (compilation.SemanticModelProvider == null)
```
Accepting any model here seems like it could easily lead to regressions in the future. Consider a few months from now if another part of the stack decides to add a light weight `SemanticModelProvider` for whatever reason. That would silently cause this part of the stack to regress behavior because it would cease to add `CachingSemanticModelProvider` instances. 

Did you consider enforcing that this requires having `CachingSemanticModelProvider`?

mavasani replied:
This is an existing public API, so we can't enforce this unfortunately. This is just a best effort optimization when this method is invoked with large number of diagnostics in same file.

jaredpar commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalyzerAction.cs:
```diff
@@ -48,161 +48,137 @@ public SymbolStartAnalyzerAction(Action<SymbolStartAnalysisContext> action, Symb
     internal sealed class SymbolEndAnalyzerAction : AnalyzerAction
     {
-        private readonly Action<SymbolAnalysisContext> _action;
+        public Action<SymbolAnalysisContext> Action { get; }
```
How did you avoid using `new` here? That should be required ...

jaredpar commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1070,32 +1070,26 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
-                    if (body != null)
+                    // If compilation has a caching semantic model provider, then cache the already-computed bound tree
+                    // onto the semantic model and store it on the event.
+                    SyntaxTreeSemanticModel semanticModelWithCachedBoundNodes = null;
```
Why did we move from a lazy to prompt get of the semantic model here? Seems like we could have created the lazy and used the `CachingSemanticModel` inside to create the model. Is there a reason that is not desirable here?

mavasani replied:
I wasn't sure if laziness would give us any benefit now that we have switched from weakly referenced semantic model in the event to a strongly referenced one that is backed by client owned caching semantic model provider. I can switch to lazy one and measure and perf changes if that helps, but code wise this looked cleaner to me.

mavasani commented:
> In the future for changes like this I really think you should consider splitting out the nullable annotations and implementation changes into different PRs. It would make it significantly easier to review.

@jaredpar Probably this got lost in the big PR description, but I did separate this out into different commits and also have a separate PR out for just the nullable changes if that is preferred for review:

>**Strongly recommended to review commit-by-commit**
1. Commit 1 (https://github.com/dotnet/roslyn/commit/f06d87c7c62a8785af74497c119e55fd0b64ef14): Nullable enable analyzer driver. I also have a separate PR out for this commit, if that is easier to review: https://github.com/dotnet/roslyn/pull/46409

2. Commit 2 (https://github.com/dotnet/roslyn/commit/039c424a0a582b4730729f40d5824e7220967087): Semantic model provider implementation for the performance improvements

jaredpar commented:
> Probably this got lost in the big PR description, but I did separate this out into different commits and also have a separate PR out for just the nullable changes if that is preferred for rev

Yep. That's on me. I missed that in the desciption 😦 

Thanks for taking the time to structure the change that way.

mavasani commented:
FYI: The nullable enable PR https://github.com/dotnet/roslyn/pull/46409 has now been merged and this branch rebased to only have the core analyzer perf improvement changes. Tagging @333fred @sharwell

333fred commented:
Will take a look now, thanks. I was waiting on the nullable changes.

333fred commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -501,10 +524,15 @@ public bool ContainsSyntaxTree(SyntaxTree syntaxTree)
         /// The event queue that this compilation was created with.
         /// </summary>
-        internal readonly AsyncQueue<CompilationEvent>? EventQueue;
+        internal AsyncQueue<CompilationEvent>? EventQueue { get; }
```
Why did this need to be changed in this PR?

mavasani replied:
So I first ended up copy-pasting this field to add the `SemanticModelProvider` and @jaredpar correctly pointed out it could be a readonly property instead - I decided to change both the existing field and newly added ones to properties.

333fred commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:
```diff
@@ -1067,15 +1074,14 @@ public ImmutableArray<Diagnostic> DequeueNonLocalDiagnosticsAndApplySuppressions
 
         private ImmutableArray<Diagnostic> FilterDiagnosticsSuppressedInSourceOrByAnalyzers(ImmutableArray<Diagnostic> diagnostics, Compilation compilation)
         {
-            diagnostics = FilterDiagnosticsSuppressedInSource(diagnostics, compilation, CurrentCompilationData.SuppressMessageAttributeState, GetOrCreateSemanticModel);
```
>GetOrCreateSemanticModel [](start = 142, length = 24)

This makes me so happy :)

333fred commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs:
```diff
@@ -0,0 +1,50 @@
+    /// For example, <see cref="CompilationWithAnalyzers"/> uses this provider to ensure that semantic model instances
+    /// are shared between the compiler and analyzers for improved analyzer execution performance. The underlying
+    /// <see cref="AnalyzerDriver"/> executing analyzers clears entries in the cache whenever a <see cref="CompilationUnitCompletedEvent"/>
+    /// has been processed, indicating all relevant analyzers have executed on the corresponding syntax tree for the event.
```
Is there a corresponding event for "All trees have finished processing"? That would allow us to ensure that the entire cache is flushed at that point as an invariant.

mavasani replied:
Yes, there is a [CompilationCompleted](http://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationCompletedEvent.cs,a37b5031cc734702,references), but I don't think we can assert the invariant as we could have inefficiently written analyzers that can invoke `compilation.GetSemanticModel` on random trees in the compilation at any time. This would be something that needs the analyzer to be fixed to avoid holding onto the model for the lifetime of compilation, but I am wondering if we should convert the strong reference via a Dictionary here instead be a weak reference via ConditionalWeakTable. I had measured performance for both these data structures on the compiler benchmark and did not see any noticeable difference so that would also be a feasible approach.

333fred replied:
Hmm. A concern with that type of analyzer: such an analyzer could request a model for a tree that has already been processed, and since the compilation is aware of the cache, it'll be added. What then clears that model from the cache? Unless I'm misunderstanding something, this is likely a memory leak. We could solve that with a CWT, but what would be the key of the CWT in this instance? The syntax tree? Or a `ConditionalWeakTable<Compilation, ConditionalWeakTable<SyntaxTree, SemanticModel>>`? It really does feel like the best solution would be, on throwing away a compilation, we throw away any cached models associated with that compilation.

333fred replied:
It feels like we either need to have the CWT of CWT solution, or more strongly tie a provider to only being used for a single compilation, otherwise we'll run into memory leak issues.

mavasani replied:
> It feels like we either need to have the CWT of CWT solution, or more strongly tie a provider to only being used for a single compilation, otherwise we'll run into memory leak issues.

Agreed. I like the latter approach as we already take a similar approach for event queue, and we already ensure that we use a new caching semantic provider for each CompilationWithAnalyzers instance, i.e. it is tied to a specific compilation instance. I will make that change.

mavasani replied:
I put a bit more thought into this, and tying a provider instance to compilation likely prevents it from ever being made public, similar to event queue on the compilation. I went with a hybrid route of `CWT<Compilation, ConcurrentDictionary<Tree, Model>>`, with the host appropriately clearing per-tree entries in the dictionary value on `CompilationUnitCompletedEvent` and clearing the CWT entry on `CompilationCompletedEvent`. I will measure and report perf numbers on the benchmark with this change.

mavasani replied:
I validated the perf numbers on the benchmark with the last commit and numbers seem similar to the ones with original implementation:

```
|                      Method |    Mean |   Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |--------:|--------:|--------:|--------:|--------:|--------:|------------:|-----------:|----------:|----------:|
| GetDiagnosticsWithAnalyzers | **12.63 s** | 0.066 s | 0.055 s | 12.61 s | 12.56 s | 12.76 s | 208000.0000 | 65000.0000 | 1000.0000 |   **1.23 GB** |
```

333fred commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs:
```diff
@@ -16,83 +18,26 @@ internal sealed class SymbolDeclaredCompilationEvent : CompilationEvent
-            _semanticModel = newSemanticModel;
+            Symbol = symbol;
+            SemanticModelWithCachedBoundNodes = semanticModelWithCachedBoundNodes;
+            _lazyCachedDeclaringReferences = new Lazy<ImmutableArray<SyntaxReference>>(() => symbol.DeclaringSyntaxReferences);
```
>_lazyCachedDeclaringReferences [](start = 12, length = 30)

Consider making this a lazily-initialized ImmutableArray with `ImmutableInterlocked.Initialize`, instead of using a `Lazy<T>`.

mavasani replied:
This is existing code, so I was trying to not change it in this PR.

333fred replied:
Consider either doing it inline, or filling an issue to follow up. We already do perf-sensitive work in this class, and this is a very-commonly-allocated class type.

---
In reply to: [469652219](https://github.com/dotnet/roslyn/pull/46508#discussion_r469652219) [](ancestors = 469652219)

333fred commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2085,9 +2119,12 @@ public new SemanticModel GetSemanticModel(SyntaxTree syntaxTree, bool ignoreAcce
+            return SemanticModelProvider?.GetSemanticModel(syntaxTree, this, ignoreAccessibility) ?? CreateSemanticModel(syntaxTree, ignoreAccessibility);
         }
 
+        internal override SemanticModel CreateSemanticModel(SyntaxTree syntaxTree, bool ignoreAccessibility)
```
Should this API also perform the same argument validation as the `Get` version?

mavasani replied:
I don't think it is required as all callers are internal and callers from public APIs would have already done the validation.

333fred commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1073,32 +1073,26 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
-                        });
+                        var syntax = body.Syntax;
+                        semanticModelWithCachedBoundNodes = (SyntaxTreeSemanticModel)cachingSemanticModelProvider.GetSemanticModel(syntax.SyntaxTree, _compilation);
+                        semanticModelWithCachedBoundNodes.GetOrAddModel(semanticModelSyntax,
```
If we move to a CWT, that might make this optimization no longer meaningful, because that semantic model could be cleaned up by the time it's used? Particularly since it's a newly-created Gen 0 object.

mavasani replied:
We will strongly hold onto this semantic model in the compilation event itself and the analyzer driver first attempts to get `SymbolDeclaredCompilationEvent.SemanticModelWithCachedBoundNodes` before attempting to fetch it from the semantic model provider. So, we are guaranteed to re-use semantic model for that event's callback.

333fred commented on src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs:
```diff
@@ -3747,5 +3748,38 @@ static ImmutableArray<Diagnostic> getReportedDiagnostics(AnalysisResult analysis
+
+            public void VerifyCachedModel(SyntaxTree tree, SemanticModel model)
+            {
+                Assert.Equal(model, _cache[tree]);
```
>Equal [](start = 23, length = 5)

Same?

333fred commented on src/Compilers/VisualBasic/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.vb:
```diff
@@ -1598,5 +1599,38 @@ End Namespace
+            End Function
+
+            Public Sub VerifyCachedModel(tree As SyntaxTree, model As SemanticModel)
+                Assert.Equal(model, _cache(tree))
```
>Equal [](start = 23, length = 5)

Same?

333fred commented:
Done review pass (commit 15)

333fred commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs:
```diff
@@ -0,0 +1,81 @@
+    internal sealed class CachingSemanticModelProvider : SemanticModelProvider
+    {
+        private readonly ConditionalWeakTable<Compilation, PerCompilationProvider> _providerCache;
+        private readonly ConditionalWeakTable<Compilation, PerCompilationProvider>.CreateValueCallback _createProviderCallback;
```
Any reason this couldn't be shared among all instances? It doesn't look like the callback actually accesses current state?

333fred replied:
I think we could probably just mark the lambda `static` and inline it where it's used, and everything should just be cached for free.

---
In reply to: [470107965](https://github.com/dotnet/roslyn/pull/46508#discussion_r470107965) [](ancestors = 470107965)

mavasani replied:
I am not comfortable with a static CWT keyed on compilation. In general any static CWT can be evil. Given that allocating this delegate once per instance of the provider is not expensive, I will retain this logic.

mavasani replied:
Oh sorry, I did not see you comment was for just the callback, not the cache. I can definitely make that static.

jaredpar commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2122,15 +2119,12 @@ internal override SemanticModel GetSemanticModelCore(SyntaxTree syntaxTree, bool
-            }
-
-            return new SyntaxTreeSemanticModel(this, syntaxTree, ignoreAccessibility);
+            return SemanticModelProvider?.GetSemanticModel(syntaxTree, this, ignoreAccessibility) ?? CreateSemanticModel(syntaxTree, ignoreAccessibility);
```
This code pattern implies that it's okay for `GetSemanticModel` to return `null` yet the declaration has it as a non-null return value. Think we should either change the definition to return `SemanticModel?` or change this code to check for `null` and throw / assert / etc ... if it's seen.

333fred replied:
Yeah, `SemanticModelProvider.GetSemanticModel` isn't annotated, so it shouldn't be returning null.

---
In reply to: [470112360](https://github.com/dotnet/roslyn/pull/46508#discussion_r470112360) [](ancestors = 470112360)

333fred commented:
Done review pass (commit 16). One minor comment about delegate caching, but this approach with the CWT looks very promising and I'll be happy to have it :)

mavasani commented:
@jaredpar Any further feedback?

mavasani commented:
Thanks for the reviews!
